recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch

```diff
@@ -0,0 +1,534 @@
++		return;
++	}
++	dm_gpio_set_value(&gpio, value);
++}
++
++static bool board_is_m2(void)
++{
++	struct iot2050_info *info = IOT2050_INFO_DATA;
++
++	return IS_ENABLED(CONFIG_TARGET_IOT2050_A53_PG2) &&
++		info->magic == IOT2050_INFO_MAGIC &&
++		strcmp((char *)info->name, "IOT2050-ADVANCED-M2") == 0;
++}
++
+ static bool board_is_advanced(void)
+ {
+ 	struct iot2050_info *info = IOT2050_INFO_DATA;
+@@ -103,6 +213,8 @@ void set_board_info_env(void)
+ 	if (board_is_advanced()) {
+ 		if (IS_ENABLED(CONFIG_TARGET_IOT2050_A53_PG1))
+ 			fdtfile = "ti/k3-am6548-iot2050-advanced.dtb";
++		else if(board_is_m2())
++			fdtfile = "ti/k3-am6548-iot2050-advanced-m2.dtb";
+ 		else
+ 			fdtfile = "ti/k3-am6548-iot2050-advanced-pg2.dtb";
+ 	} else {
+@@ -118,6 +230,103 @@ void set_board_info_env(void)
+ 	env_save();
+ }
+ 
++static void m2_overlay_prepare(void)
++{
++#if defined(CONFIG_OF_LIBFDT) && defined(CONFIG_OF_BOARD_SETUP)
++	const char *overlay_path;
++	void *overlay;
++	u64 loadaddr;
++	ofnode node;
++	int ret;
++
++	if (connector_mode == BKEY_PCIEX2_EKEY_NONE)
++		return;
++
++	if (connector_mode == BKEY_PCIE_EKEY_PCIE)
++		overlay_path = "/fit-images/bkey-ekey-pcie-overlay";
++	else
++		overlay_path = "/fit-images/bkey-usb3-overlay";
++
++	node = ofnode_path(overlay_path);
++	if (!ofnode_valid(node))
++		goto fit_error;
++
++	ret = ofnode_read_u64(node, "load", &loadaddr);
++	if (ret)
++		goto fit_error;
++
++	ret = ofnode_read_u32(node, "size", &connector_overlay_size);
++	if (ret)
++		goto fit_error;
++
++	overlay = map_sysmem(loadaddr, connector_overlay_size);
++
++	connector_overlay = malloc(connector_overlay_size);
++	if (!connector_overlay)
++		goto fit_error;
++
++	memcpy(connector_overlay, overlay, connector_overlay_size);
++	return;
++
++fit_error:
++	pr_err("M.2 device tree overlay %s not available,\n", overlay_path);
++#endif
++}
++
++static void m2_connector_setup(void)
++{
++	ulong m2_manual_config = env_get_ulong("m2_manual_config", 10,
++					       CONNECTOR_MODE_INVALID);
++	const char *mode_info = "";
++	struct m2_config_pins config_pins;
++	unsigned int n;
++
++	/* enable M.2 connector power */
++	set_pinvalue("gpio@601000_17", "P3V3_M2_EN", 1);
++	udelay(4 * 100);
++
++	m2_manual_config = env_get_ulong("m2_manual_config", 10,
++					 CONNECTOR_MODE_INVALID);
++	if (m2_manual_config < CONNECTOR_MODE_INVALID) {
++		mode_info = " [manual mode]";
++		connector_mode = m2_manual_config;
++	} else { /* auto detection */
++		for (n = 0; n < ARRAY_SIZE(config_pins.config); n++)
++			config_pins.config[n] =
++				get_pinvalue(m2_bkey_cfg_pin_info[n].gpio_name,
++					     m2_bkey_cfg_pin_info[n].label);
++		connector_mode = CONNECTOR_MODE_INVALID;
++		for (n = 0; n < ARRAY_SIZE(m2_config_table); n++) {
++			if (!memcmp(config_pins.config,
++				    m2_config_table[n].config_pins.config,
++				    sizeof(config_pins.config))) {
++				connector_mode = m2_config_table[n].mode;
```

Comment:
mode_info should also be set

Reply:
Empty mode_info implies "auto" so far. I don't think we need to write this out explicitly.

Reply:
We write out the detected mode already.